elastic/kibana
# [Security solution] AI Assistant Telemetry for Knowledge Base

## Summary

AI Assistant Telemetry for Knowledge Base and RAG Alerts

### Renaming

Updated variable names for accuracy and clarity.
1. Renames `assistantLangChain` boolean to `isEnabledKnowledgeBase`. (@elastic/response-ops is tagged for review because of this renaming)
2. Renames `alerts` boolean to `isEnabledRAGAlerts`.

### Telemetry

1. Enhances existing EBT to include new fields: `Assistant Message Sent`
    - Adds the following properties to each event: `isEnabledKnowledgeBase` & `isEnabledRAGAlerts`
3. Adds new client side EBT to track settings changes: `Assistant Setting Toggled`
    - Events contain one or both of the following properties: `isEnabledKnowledgeBase` & `isEnabledRAGAlerts`
4. Adds server side EBT
    - `knowledge_base_execution_success` - successful returned ES similaritySearch using knowledge base. Tracks which knowledge base resource was accessed, which ELSER model used to execute the knowledge base query, the number of documents returned from ES, and the time it took for ES to resolve
    - `knowledge_base_execution_error` - ES similaritySearch error, includes error message, knowledge base resource was accessed, and ELSER model 
    - `invoke_assistant_success`- Success invoking assistant, includes error message, and booleans `isEnabledKnowledgeBase` & `isEnabledRAGAlerts`
    - `invoke_assistant_error`- Error invoking assistant, includes error message, and booleans `isEnabledKnowledgeBase` & `isEnabledRAGAlerts`

## Dashboards

I've added to the [existing GenAis Dashboard](https://p.elstc.co/paste/hZDHUe4q#jWAd6xHZZsAlDd2rUzhuUrnuw6tn-dtYU1t+XNqJ+0Y) to show off this new telemetry. The visualizations from `KB and RAG on Alerts per message sent` on down are all new. 


## Testing

Below are testing steps along with the events they should trigger. Please confirm the events below each step in the console and on the telemetry staging server. Remember the client events will be in `ebt-kibana-browser` and server will be `ebt-kibana-server`. It seems like the browser data view updates more frequently than the server `¯\_(ツ)_/¯`

1. Ensure `telemetry.optIn: true` is defined in your dev yml
2. Have a working GenAI connector and a broken one (bad credentials or model)
3. Toggle on the knowledge base and save (RAG Alerts should be off) 
    - `Assistant Setting Toggled - { isEnabledKnowledgeBase: true }`
4. Send a message with the working connector 
    - `invoke_assistant_success - { isEnabledKnowledgeBase: true, isEnabledRAGAlerts: false }`
   -  `knowledge_base_execution_success - { "model": ".elser_model_2", "resourceAccessed": "esql", "resultCount": 52, "responseTime": 35 }`
   - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":true,"role":"user","conversationId":"Event summary"}`
   - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":true,"role":"assistant","conversationId":"Event summary"}`
5. Send a message with the broken connector 
    - `invoke_assistant_error - { errorMessage: 'something', isEnabledKnowledgeBase: true, isEnabledRAGAlerts: false }`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":true,"role":"user","conversationId":"Event summary"}`
   - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":true,"role":"assistant","conversationId":"Event summary"}`
6. I wasn't sure how to get an actual error from knowledge base, so to test I did `throw new Error` in `similaritySearch`. To get it to throw, send a message with the working connector  with KB on
    - `knowledge_base_execution_error - { "errorMessage": "something" , "model": ".elser_model_2", "resourceAccessed": "esql"}`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":true,"role":"user","conversationId":"Event summary"}`
   - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":true,"role":"assistant","conversationId":"Event summary"}`
7. Remove the error you threw in the last test
8. Toggle on RAG Alerts, leaving knowledge base on
    - `Assistant Setting Toggled - { isEnabledRAGAlerts: true }`
9. Send a message with the working connector 
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":true,"role":"user","conversationId":"Event summary"}`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":true,"role":"assistant","conversationId":"Event summary"}`
    - `invoke_assistant_success - { isEnabledKnowledgeBase: true, isEnabledRAGAlerts: true }`
    - `knowledge_base_execution_success - { "model": ".elser_model_2", "resourceAccessed": "esql", "resultCount": 52, "responseTime": 35 }`
10. Send a message with the broken connector 
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":true,"role":"user","conversationId":"Event summary"}`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":true,"role":"assistant","conversationId":"Event summary"}`
    - `invoke_assistant_error - { errorMessage: 'something', isEnabledKnowledgeBase: true, isEnabledRAGAlerts: true }`
11. Toggle off both RAG Alerts and knowledge base
    - `Assistant Setting Toggled - {  isEnabledKnowledgeBase: false, isEnabledRAGAlerts: false }`
12. Send a message with the working connector 
    - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":false,"role":"user","conversationId":"Event summary"}`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":false,"role":"assistant","conversationId":"Event summary"}`
    - `invoke_assistant_success - { isEnabledKnowledgeBase: false, isEnabledRAGAlerts: false }`
13. Send a message with the broken connector 
    - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":false,"role":"user","conversationId":"Event summary"}`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":false,"isEnabledKnowledgeBase":false,"role":"assistant","conversationId":"Event summary"}`
    - `invoke_assistant_error - { errorMessage: 'something', isEnabledKnowledgeBase: false, isEnabledRAGAlerts: false }`
14. Toggle on only RAG Alerts, leaving knowledge base off
    - `Assistant Setting Toggled - { isEnabledRAGAlerts: true }`
15. Send a message with the working connector 
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":false,"role":"user","conversationId":"Event summary"}`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":false,"role":"assistant","conversationId":"Event summary"}`
    - `invoke_assistant_success - { isEnabledKnowledgeBase: false, isEnabledRAGAlerts: true }`
16. Send a message with the broken connector 
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":false,"role":"user","conversationId":"Event summary"}`
    - `Assistant Message Sent - {"isEnabledRAGAlerts":true,"isEnabledKnowledgeBase":false,"role":"assistant","conversationId":"Event summary"}`
    - `invoke_assistant_error - { errorMessage: 'something', isEnabledKnowledgeBase: false, isEnabledRAGAlerts: true }`
